### PR TITLE
Fix typography when text is empty 

### PR DESCRIPTION
**Pull request by vol24pl**
# What was fixed
- Add gitignore
- In `LineHeightViewController` move text after typography changes to see if the issue is fixed.
- Make observing a higher order function instead of using hardcoded closure.
- Add static dictionary where key is the UILabel and value are the attributes for this label. This is a hack to have a stored property in extension without using objc.
- Whenever attribute is fetched via getter, check if it should use the static dictionary value. If so, swap the values in label and remove the values from the dictionary. Also fix the text alignment that would otherwise be reset.
- Whenever attribute is added to empty string, store it in the static dictionary for a specific label.
- Whenever attribute is removed, remove it in the dictionary as well.
- Make closure that is triggered on text change do two things
  - a) _ = self.attributedText -> This code should not do anything as it's only getter from the UIKit, but internal implementation of the UIKit has side-effects that are actually fixing the typography problems! It does not have to be a setter as in previous implementation.
  - b) _ = self.attributes -> This code is also a getter, but this logic needs to be triggered each time text changes to swap values if previous static dictionary had values (hence, previous value had to be empty string).
NOTE: Code from line "a" is not necessary, as code from line "b" contains code from line "a" anyway. It's added here for the brevity of the iOS Framework side effects

# Tests
Following changes for UILabel were tested manually. **All tested passed**
```
    // lineHeight = 50 -> "BBB"
    // lineHeight = 50 -> "BBB" -> "CCC"
    // lineHeight = 50 -> "" -> "BBB"
    // lineHeight = 50 -> "BBB" -> ""-> "CCC"
    // lineHeight = 50 -> "BBB" -> ""-> "" -> "CCC"
    // lineHeight = 50 -> "" -> ""-> "" -> "AAA"
    // lineHeight = 50 -> "" -> "" -> lineHeight = nil -> "" -> "BBB"
    // lineHeight = 50 -> "BBB" -> "" -> lineHeight = nil -> "" -> "CCC"
    // lineHeight = 50 -> "" -> lineHeight = nil -> lineHeight = 50 -> "BBB"
    // lineHeight = 50 -> "" -> lineHeight = nil -> "BBB"


    // "" -> lineHeight = 50 -> "BBB"
    // "" -> lineHeight = 50 -> "BBB" -> "CCC"
    // "" -> lineHeight = 50 -> "" -> "BBB" 
    // "" -> lineHeight = 50 -> "BBB" -> ""-> "CCC"
    // "" -> lineHeight = 50 -> "BBB" -> ""-> "" -> "CCC"
    // "" -> lineHeight = 50 -> "" -> ""-> "" -> "AAA"
    // "" -> lineHeight = 50 -> "" -> lineHeight = nil -> "BBB"
    // "" -> lineHeight = 50 -> "" -> lineHeight = nil -> lineHeight = 50 -> "BBB"
    // "" -> lineHeight = 50 -> "" -> "" -> lineHeight = nil -> "" -> "BBB"
    // "" -> lineHeight = 50 -> "BBB" -> "" -> lineHeight = nil -> "" -> "CCC"

    // "AAA" -> lineHeight = 50 -> "BBB"
    // "AAA" -> lineHeight = 50 -> "BBB" -> "CCC"
    // "AAA" -> lineHeight = 50 -> "" -> "BBB"
    // "AAA" -> lineHeight = 50 -> "BBB" -> ""-> "CCC"
    // "AAA" -> lineHeight = 50 -> "BBB" -> ""-> "" -> "CCC"
    // "AAA" -> lineHeight = 50 -> "" -> ""-> "" -> "AAA"
    // "AAA" -> lineHeight = 50 -> "" -> lineHeight = nil -> "BBB"
    // "AAA" -> lineHeight = 50 -> "" -> lineHeight = nil -> lineHeight = 50 -> "BBB"
    // "AAA" -> lineHeight = 50 -> "BBB" -> "" -> lineHeight = nil -> "" -> "CCC"
    // "AAA" -> lineHeight = 50 -> "" -> "" -> lineHeight = nil -> "" -> "BBB"
```

_Issue:_  https://github.com/Geri-Borbas/iOS.Blog.UILabel_Typography_Extensions/issues/1

### CODE REVIEW

**Comment by vol24pl** [comment]
(force pushed to use correct GPG keys to my commits)

**Comment by Geri-Borbas** [comment]
Thanks for the changeset, and the thorough documentation! 🙌 I fetched the fork, looking at it right now.

First impression: I'm not sure about keeping a static strong reference to every `UILabel` (as a dictionary key) that opts in to use this, as it seems (only by reading the code at least) in many cases it will leak the label for eternity (even after the parent screen goes away).

**Comment by vol24pl** [comment]
You are right. I need to think if there's a way to clean the dict entry once the `UILabel` is deinited! 

This will leak only if the UILabel is created with `""`, style is added and UILabel is destroyed without ever changing text from empty string to anything else. So not that big of a deal, but still a leak! I missed that!

Also this is used more or less for the same effect as the 

```swift
fileprivate var observer: TextObserver? {
		get {
			objc_getAssociatedObject(self, &Keys.observer) as? TextObserver
		}
		set {
			objc_setAssociatedObject(self, &Keys.observer, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
		}
	}
```

This is just an alternative approach for storing state inside the extension. We could stick to one way of doing that, if you prefer :)

===

Once you decide what do you want to do with this PR/Code and if you want to marge it, I'll create a 2nd PR that refactors UILabel extension into a generic protocol+extension with default implementation that works for both `UILabel` and `UITextView`, as I have this code already made for my project :)

**Comment by Geri-Borbas** [comment]
I don't think `deinit` will ever be called this way. As long as the label is retained as the key of the static dictionary on `UILabel`.

I think the associated object approach is far safer, considering it works the other way around. Everything that is associated with a given label instance gets released as long as the label goes away, **without ever affecting the label's lifecycle**.

Putting it to a distinct protocol extension sounds super great, as it nicely separates the code concerned with working around the empty string issue.

Thanks for the contribution! I'm sure happy to merge (once we iron out the details above), it benefits everyone dealing with this annoying empty string behavior.

> It takes a bit more time than usual, as I review all the considerations/alternatives one by one, as I see the blog-related repositories as part of my portfolio.

**Comment by vol24pl** [comment]
`UILabel` was NOT retained. Dictionary does not store the keys. It only uses the keys to evaluate hash from them. Dictionary stores the value in memory cell corresponding with hash value - the key (in naive implementation, in real life there are hash buckets, etc.). That's why it's complexity to add value is O(1). It does not iterate, it just checks memory cell correspinding with the hash and uses it.

So `deinit` of `UILabel` would work as usual, but value (in this case `[NSAttributedString.Key: Any]`) was retained and never left.

To make this work we'd need to do something like this:

```swift
class NewLabel: UILabel {
    deinit {
        UILabel.dictionary[self] = nil // removing the value of type [NSAttributedString.Key: Any]
    }
}
```

But override would require class to inherit from `UILabel` and the whole point of your solution was to avoid inheritance/composition :)

Anyway, memory leak fixed.

**Comment by vol24pl** [comment]
@Geri-Borbas did you have a time to review it?

**Comment by vol24pl** [comment]
@Geri-Borbas 
image

**Comment by Geri-Borbas** [comment]
Thanks for that again. In the upcoming weeks I'll review/finish up the corresponding article as well, so I'll incorporate this work as well.

**Comment by vol24pl** [comment]
Ok. Keep in mind that I've got prepared code for another PR (after you merged this one) that makes this solution work with not only `UILabel` but also `UITextView` and `UITextField` :)

**Comment by vol24pl** [comment]
Friendly reminder. How is the article going? Can I help with some more details?

**Comment by Geri-Borbas** [comment]
Actually, this one became the number one article on my blog by far. Now I feel 2x obligated to handle empty strings. 😅 Top of the list once I act on the blog.

File: UILabel_Typography_Extensions/Typography/UILabel+Observer.swift
Comment (Geri-Borbas) [inline UILabel_Typography_Extensions/Typography/UILabel+Observer.swift:47]: Thanks, makes a lot of sense to move out the use-case from the extension, leaving it concerned with observing only. 🙌

Comment (Geri-Borbas) [reply UILabel_Typography_Extensions/Typography/UILabel+Observer.swift:47]: Gonna bubble this up to the article as well. 🤔

File: UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift
Comment (Geri-Borbas) [inline UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift:43]: This is so cool if simply pinging the getter has the same effect, nice catch! 😆

File: UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift
Comment (Geri-Borbas) [inline UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift:91]: I'm thinking of a more simplified solution for this, simply storing an `NSAttributedString` that holds the attributes even when the label is empty, not sure.

**Review by Geri-Borbas — APPROVED** [review]
I'm thinking of some modifications before merging the feature into `main`, also I need to update the article with the corresponding code pieces.

File: UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift
Comment (Geri-Borbas) [reply UILabel_Typography_Extensions/Typography/UILabel+Typograpy.swift:91]: UPDATE: I ended up doing just that, also (heavily) updated the article at http://blog.eppz.eu/uilabel-line-height-letter-spacing-and-more-uilabel-typography-extensions/ with all the changes. Thanks again for the inputs! 💎✨

**Pull request merged**
